### PR TITLE
feat: add search retention to ResourceBrowserFilter

### PR DESCRIPTION
**Pull request by Elessar1802**
# Description

The search on any k8sResource of resource browser should be retained upon page refresh.

Fixes https://github.com/devtron-labs/devtron/issues/4774

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/ResourceBrowser/ResourceList/ResourceFilterOptions.tsx [hunk 1]
```diff
@@ -70,7 +71,10 @@ const ResourceFilterOptions = ({
     }
 
     const handleOnChangeSearchText = (event): void => {
-        setSearchText(event.target.value)
+        const text = event.target.value
+        setSearchText(text)
+        const queryStr = updateQueryString(location, [['search', text]])
```
Comment (Elessar1802) [inline src/components/ResourceBrowser/ResourceList/ResourceFilterOptions.tsx:76]: 'search' could be constant

File: src/components/ResourceBrowser/ResourceList/ResourceList.tsx [hunk 2]
```diff
@@ -315,7 +319,6 @@ export default function ResourceList() {
             !isNodes
         ) {
             selectedResource.gvk.Kind !== SIDEBAR_KEYS.nodeGVK.Kind && getResourceListData()
-            setSearchText('')
```
Comment (Elessar1802) [inline src/components/ResourceBrowser/ResourceList/ResourceList.tsx:318]: Need to remove this cuz on a page refresh `SelectedResource` changes from `null -> Node -> (finally correct type)`

File: src/components/ResourceBrowser/ResourceList/ResourceList.tsx [hunk 3]
```diff
@@ -984,6 +987,7 @@ export default function ResourceList() {
                     updateResourceSelectionData={updateResourceSelectionData}
                     isCreateModalOpen={showCreateResourceModal}
                     isClusterError={!!clusterErrorTitle}
+                    setResourceListText={setSearchText}
```
Comment (Elessar1802) [inline src/components/ResourceBrowser/ResourceList/ResourceList.tsx:990]: maybe the name could be better?

**Review by vivek-devtron — APPROVED** [review]

File: src/components/ResourceBrowser/Utils.ts [hunk 4]
```diff
@@ -288,3 +290,23 @@ export const convertK8sObjectMapToOptionsList = (
 
     return _k8sObjectOptionsList
 }
+
+export const updateQueryString = (
+    location: ReturnType<typeof useLocation>,
+    entries: [key: string, value: string][],
+): string => {
+    const qs = queryString.parse(location.search)
```
Comment (AbhishekA1509) [inline src/components/ResourceBrowser/Utils.ts:298]: name of qs can be better

File: src/components/ResourceBrowser/Utils.ts [hunk 5]
```diff
@@ -288,3 +290,23 @@ export const convertK8sObjectMapToOptionsList = (
 
     return _k8sObjectOptionsList
 }
+
+export const updateQueryString = (
+    location: ReturnType<typeof useLocation>,
+    entries: [key: string, value: string][],
```
Comment (AbhishekA1509) [inline src/components/ResourceBrowser/Utils.ts:296]: can we import it as a interface

File: src/components/ResourceBrowser/Utils.ts [hunk 6]
```diff
@@ -288,3 +290,23 @@ export const convertK8sObjectMapToOptionsList = (
...
+): string => {
+    const qs = queryString.parse(location.search)
+    const keys = Object.keys(qs)
+    const query = {}
+    keys.forEach((key) => {
+        query[key] = qs[key]
+    })
+    for (const [key, value] of entries) {
```
Comment (AbhishekA1509) [inline src/components/ResourceBrowser/Utils.ts:304]: should try to avoid for loop

**Pull request merged**